local-volume/utils/update-pv-to-beta/update_pv_na_to_beta.go

```diff
@@ -0,0 +1,132 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/kubernetes-incubator/external-storage/local-volume/provisioner/pkg/common"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+func cloneAndUpdatePV(pv *v1.PersistentVolume) (*v1.PersistentVolume, error) {
+	var affinity v1.NodeAffinity
+	err := json.Unmarshal([]byte(pv.Annotations[v1.AlphaStorageNodeAffinityAnnotation]), &affinity)
+	if err != nil {
+		return nil, err
+	}
+	pvClone := pv.DeepCopy()
+	delete(pvClone.Annotations, v1.AlphaStorageNodeAffinityAnnotation)
+	pvClone.Spec.NodeAffinity = &v1.VolumeNodeAffinity{
+		Required: affinity.RequiredDuringSchedulingIgnoredDuringExecution,
+	}
+	return pvClone, nil
+}
+func updateLocalPVAlphaAnn(client *kubernetes.Clientset, pv *v1.PersistentVolume) error {
+	// if it is not local PV, return directly
+	if pv.Spec.Local == nil {
+		return nil
+	}
+	// check alpha node affinity annotation
+	if len(pv.Annotations) > 0 && pv.Annotations[v1.AlphaStorageNodeAffinityAnnotation] != "" {
+		pvClone, err := cloneAndUpdatePV(pv)
+		if err != nil {
+			return err
+		}
+
+		glog.Infof("Updating local PV(%s), node affinity is: %v", pv.Name, pv.Annotations[v1.AlphaStorageNodeAffinityAnnotation])
+		_, err = client.CoreV1().PersistentVolumes().Update(pvClone)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				glog.Errorf("PV(%s) seems to have been deleted", pv.Name)
+				return nil
+			}
+			if errors.IsConflict(err) {
+				glog.Errorf("PV(%s) seems to have been updated, will get the latest resource version and retry", pv.Name)
+				return getAndUpdatePV(pv.Name, client)
```

Comment:
Is this needed?  The next iteration of startSwitching will handle this case right?

Reply:
@msau42 This was my suggestion. In a long list and update cycle, this error is likely to happen. Is this tool expected to run periodically?

Reply:
What if we change this to keep on running forever until there are no more PVs to update.

Reply:
Yes, that would also work. I think that couple of retries and then giving up with the list of failed PVs would be sufficient.

Reply:
It is hard to `keep on running forever until there are no more PVs to update.` . What if new local PVs to be added after we did `list` operation.  Unless we keep trying and never exit

Reply:
After thinking about it more, Jobs should restart the pod until the pod successfully completes.  Instead of doing retries inside the container, can we just rely on Jobs to keep on retrying until success?

Reply:
You mean setting the `.spec.template.spec.restartPolicy ` to `OnFailure`.
Will this keep `list`ing as soon as it failed and increase the burden of API ? If the job failed because of `TooManyRequests`, it will make the situation worse. Maybe we can sleep some time, and then exit with error code ?